### ISSUE TITLE
get impact force when rigidbody hit an other body

### ISSUE THREAD

**Issue by Wapit1** [godotengine/godot-proposals#487]
**Describe the project you are working on:**
I have been working with physic game where the player hits a lot of body with held rigidbody
and can't get more progress on the physic side of thing because of the lack of a consistant lightweight way of getting the impact force
**Describe the problem or limitation you are having in your project:**
getting the impact force requires to constantly calculate the previous velocity, else you can't tell how fast the object hit one a other: the force is already reduce before you receive the signal and you thus get a very small number which isn't representative of the linear velocity before impact 
(due note that it 2D you have still a bit more force during the impact tick, but it is not that consistant, has it already lost a bit force )
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
having a consistant Vector of the impact strength  will allow programmer to create easily system like sound of impact, realistic breakable object, multiplied knockback on rigidbody and other 
**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
having a parameter with the signal on_body_entered would allow the user to use it easily and should be , too my understanding of physics engine, not too hard to get has the physics engine already calculates the force being transferred to the other object so the number would only need to be exposed
 (and while your at it while not expose a few more variable like force lost in friction, the average of the impact point and the force a joint is applying to the rigidbody)
**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It would be used often and can get worked around with a few line of code (keeping in a var the linear_velocity every tick), but it is required to be put on every rigidbody (has you want the force difference on impact for realistic result) and its is not that consistant, has the body might go through the object and then get pulled back
**Is there a reason why this should be core and not an add-on in the asset library?:**
because having it has an add-on would be harder to implement and most developer would like the idea of easily have the possibility to have different impact sound based on impact strength, have damage dealt based on the force of the swinging weapon and thus be allowed to create loads of physic based emergent gameplay

**Comment by Zireael07** [comment]
I have yet to see the 

> the body might go through the object and then get pulled back

case, but having a built in function would definitely make having various impact related systems much easier (e.g. deforming a visual mesh on impact like I do in my little car game)

**Comment by fire** [comment]
Is this related to https://github.com/godotengine/godot/pull/36008 ?

**Comment by Wapit1** [comment]
> Is this related to [godotengine/godot#36008](https://github.com/godotengine/godot/pull/36008) ?

no, but I am always happy to see more physics control (I am personally waiting for active ragdoll - currently working with joint and ragdoll)

**Comment by fire** [comment]
I linked the wrong pr. This is active ragdolls. https://github.com/godotengine/godot/pull/28909

**Comment by Wapit1** [comment]
yes I am waiting for this PR (although it is a separate issue)
my current project is so because of the lack of physics data and reliable active ragdoll 
having the player smash stuff with is floppy arms

**Comment by briansemrau** [comment]
Is this the same as #409? That feature request describes the ability to use RigidBody2D contact monitoring to get the collision impulse.

**Comment by Jimmio92** [comment]
If anyone else encounters this, I worked around this limitation by keeping track of the previous velocity, and on body_entered signal, using the difference between the previous _physics_process' velocity and the current one as a way to determine "force" (at least well enough to make hit sounds work) as it seems to run the signal *after* taking account for the hit

**Comment by MrMinimal** [comment]
> If anyone else encounters this, I worked around this limitation by keeping track of the previous velocity, and on body_entered signal, using the difference between the previous _physics_process' velocity and the current one as a way to determine "force" (at least well enough to make hit sounds work) as it seems to run the signal _after_ taking account for the hit

That doesn't work for rotating objects. Their impace force can be multiple times the force of just the velocity difference. Still a very valdid proposal.

**Comment by graphnode** [comment]
> That doesn't work for rotating objects. Their impace force can be multiple times the force of just the velocity difference. Still a very valdid proposal.

Could also keep track of angular velocity but...

I haven't tried but getting the body state and then do the [dot product](https://docs.godotengine.org/en/stable/tutorials/math/vector_math.html#dot-product) between [collision normal](https://docs.godotengine.org/en/stable/classes/class_physicsdirectbodystate3d.html#class-physicsdirectbodystate3d-method-get-contact-local-normal) and [collision velocity](get_contact_local_velocity_at_position) multiplied by the other collider mass might give better results.

**Comment by jitspoe** [comment]
Here's how I approached it:

```gdscript
On the rigid body:
func _integrate_forces(state : PhysicsDirectBodyState3D):
	for contact_index in state.get_contact_count():
		var object_hit := state.get_contact_collider_object(contact_index)
		if (is_instance_valid(object_hit)): # To fix a case where an object hits the player as player is deleted during level transition (intermission)
			Physics.handle_rigid_body_collision(self, object_hit, state.get_contact_collider_position(contact_index), state.get_contact_local_velocity_at_position(contact_index), state.get_contact_impulse(contact_index))

And in my physics Autoload:
static func handle_rigid_body_collision(object : PhysicsProp, object_hit : Node3D, position : Vector3, velocity : Vector3, impulse : Vector3):
	# TODO: Sometimes impulse is 0, 0, 0, even though there is a solid impact.  Jolt bug?
	if (impulse.length_squared() > MIN_IMPULSE_RESPONSE_SQUARED):
		Combat.handle_rigid_body_hit(object, object_hit, velocity, impulse)
		var speed_change := impulse.length() / object.mass
		SoundSystem.play_physics_impact_sound(object, position, speed_change)
	else:
		pass
```

Still, it would be great to have a general purpose callback with the impulse whenever a collision happened so things like impact sounds and such could be implemented without having a ton of extra script in performance-critical areas like physics.

**Comment by Calinou** [comment]
> Still, it would be great to have a general purpose callback with the impulse whenever a collision happened

How would the threshold for "a collision happened" be determined for sending this signal? I'd imagine that if you have no threshold, then tiny velocity changes after a collision would lead to a signal being emitted on every physics frame.

**Comment by zachariahchandler** [comment]
If you don't want to implement your own force integration, you can also get direct body states by RID like so (sorry it's all in C#):

```csharp
public static PhysicsDirectBodyState3D GetDirectBodyState3D (CollisionObject3D body) {
  return PhysicsServer3D.BodyGetDirectState(body.GetRid());
}
```

However, as noted in jitspoe's comment, sometimes getting the collision impulse from the direct body state returns zero, which seems to be a manifestation of [this bug](https://github.com/godotengine/godot/issues/73541), so it's much more reliable to compare velocities across frames and deduce the impulse from the difference:

```csharp
// in the rigidbody...
public Vector3 lastLinearVelocity;
public Vector3 lastAngularVelocity;
public Vector3 lastGlobalPosition;

public override void _PhysicsProcess (double delta) {
  UpdateCachedVectors();
}
  
void UpdateCachedVectors () {
  lastGlobalPosition = GlobalPosition;
  lastLinearVelocity = LinearVelocity;
  lastAngularVelocity = AngularVelocity;
}

// on body enter signal...
public void OnBodyEnter (Node body) {
  PhysicsDirectBodyState3D state = PhysicsServer3D.BodyGetDirectState(this.GetRid());
  
  for (int i = 0; i < state.GetContactCount(); i++) {
    Vector3 contactPt = state.GetContactColliderPosition(i);
  
    Vector3 currentPtVelocity = GetPointVelocity(contactPt, LinearVelocity, AngularVelocity, this);
    
    Vector3 lastPtVelocity = GetPointVelocity(contactPt, lastLinearVelocity, lastAngularVelocity, this);
    
    Vector3 impulse = currentPtVelocity - lastPtVelocity;
    // ^ this isn't a very detailed approach, and it doesn't account for like, delta or mass or anything, but you could add that if needed. in my use-case i just need to know relatively how hard something was hit
    
    if (impulse.LengthSquared() < minimumContactImpulseSquared) {
      continue;
    }
    // implicit else...

    // now you can handle the collision however you want...
  }
}

// i store this in a static utility class...
public static Vector3 GetPointVelocity (Vector3 point, Vector3 linearVelocity, Vector3 angularVelocity, Vector3 relativeBodyGlobalPosition) {
  return linearVelocity + angularVelocity.Cross(point - relativeBodyGlobalPosition);
}
```

I'm sure this method has its problems too, and it would still be better if we had a built-in method that worked (and was easily accessible in the `body_entered` signal!).
